### PR TITLE
Aggregative Nested Loop Join Operator

### PR DESCRIPTION
**Pull request by metesynnada**
## Which issue does this PR close?

Closes #.

## Rationale for this change

We want to run queries like
```sql
SELECT l_partkey, LAST_VALUE(o_orderkey ORDER BY o_orderkey) as sa
FROM orders, lineitem
WHERE l_orderkey >= o_orderkey
GROUP BY l_orderkey;
```
in a pipeline-friendly fashion.

## What changes are included in this PR?
This PR adds `AggregativeNestedLoopJoinExec`, a non-equijoin version of the `AggregativeHashJoinExec`. This improves support for join analysis based on selectivity on aggregate columns.

### CODE REVIEW

File: datafusion/physical-plan/src/joins/aggregator_nested_loop_join.rs [hunk 1]
```diff
@@ -0,0 +1,1809 @@
...
+use crate::joins::nested_loop_join::distribution_from_join_type;
+use crate::stream::RecordBatchBroadcastStreamsBuilder;
+
+use async_trait::async_trait;
+use futures::Stream;
+
+use itertools::Itertools;
+use parking_lot::Mutex;
+
+/// Represents a partitioned hash join execution plan.
```
Comment (mustafasrepo) [inline datafusion/physical-plan/src/joins/aggregator_nested_loop_join.rs:61]: This line seems like a leftover

File: datafusion/physical-plan/src/joins/aggregator_nested_loop_join.rs [hunk 2]
```diff
@@ -0,0 +1,1809 @@
...
+///
+/// The `AggregatorNestedLoopJoinExec` struct facilitates the execution of nested loop join operations in
+/// parallel across multiple partitions of data. It takes two input streams (`left` and `right`),
+/// a set of common columns to join on (`on`), and applies a join filter to find matching rows.
```
Comment (mustafasrepo) [inline datafusion/physical-plan/src/joins/aggregator_nested_loop_join.rs:65]: There is no join on field

File: datafusion/physical-plan/src/joins/aggregator_nested_loop_join.rs [hunk 3]
```diff
@@ -0,0 +1,1809 @@
...
+    ///
+    /// # Returns
+    /// * A `Result` indicating the success or failure of the pruning operation.
+    fn prune(
+        &mut self,
+        filter: &JoinFilter,
+        build_shrunk_exprs: Vec<(PhysicalSortExpr, Interval)>,
+        fetch_size: usize,
+    ) -> Result<()> {
+        // print_batches(&[self.record_batch.clone()]);
```
Comment (mustafasrepo) [inline datafusion/physical-plan/src/joins/aggregator_nested_loop_join.rs:337]:
```suggestion
```

**Comment by ozankabak** [comment]
I did a first pass and it looks good in general. I will do a deeper read-through tomorrow and we will finalize.

One small linguistic fix: Let's use the adjective `aggregative` instead of the noun `aggregator`. For example, `AggregativeNestedLoopJoin` instead of `AggregatorNestedLoopJoin` etc.

I didn't have time today to do a through check for any code reuse possibilities between SHJ and this, so please take a quick look tomorrow morning from that perspective as well.

**Review by ozankabak — APPROVED** [review]
Just finished reviewing and looks good in general. There are a few points to discuss before merging, but nothing very big. Let's do a final discussion tomorrow to finalize and then merge.

**Pull request merged**